CodeGeneration/Sources/SyntaxSupport/CompatibilityLayers.swift

```diff
@@ -0,0 +1,192 @@
+  }
+
+  private mutating func realizeLayers(for trait: Trait) {
+    guard deprecatedVarsByTrait[trait.traitName] == nil else {
+      return
+    }
+
+    let result = computeLayersFor(
+      typeName: trait.traitName,
+      initialChildren: trait.children,
+      history: trait.childHistory,
+      areRequirements: true
+    )
+
+    deprecatedVarsByTrait[trait.traitName] = result.vars
+    deprecatedInitSignaturesByTrait[trait.traitName] = result.initSignatures
+  }
+
+  /// Compute and cache compatibility layer information for the given children.
+  private mutating func computeLayersFor(
+    typeName: String,
+    initialChildren: [Child],
+    history: Child.History,
+    areRequirements: Bool
+  ) -> (vars: [Child], initSignatures: [InitSignature]) {
+    // The results that will ultimately be saved into the *ByNode dictionaries.
+    var vars: [Child] = []
+    var initSignatures: [InitSignature] = []
+
+    // Temporary working state for the loop.
+    var children = initialChildren
+    var knownVars = Set(children)
+
+    func firstIndexOfChild(named targetName: String) -> Int {
+      guard let i = children.firstIndex(where: { $0.name == targetName }) else {
+        fatalError("couldn't find '\(targetName)' in current children of \(typeName): \(String(reflecting: children.map(\.name)))")
+      }
+      return i
+    }
+
+    for changeSet in history {
+      var unexpectedChildrenWithNewNames: Set<Child> = []
+
+      // First pass: Apply the changes explicitly specified in the change set.
+      for (currentName, refactoring) in changeSet {
+        let i = firstIndexOfChild(named: currentName)
+
+        let replacementChildren = replacementChildren(for: children[i], by: refactoring)
+        children.replaceSubrange(i...i, with: replacementChildren)
+
+        if !areRequirements {
+          // Mark adjacent unexpected node children whose names have changed too.
+          if currentName != (replacementChildren.first?.name ?? "") {
+            unexpectedChildrenWithNewNames.insert(children[i - 1])
+          }
+          if currentName != (replacementChildren.last?.name ?? "") {
+            unexpectedChildrenWithNewNames.insert(children[i + replacementChildren.count])
+          }
+        }
+      }
+
+      // Second pass: Update unexpected node children adjacent to those changes whose names have probably changed.
+      for unexpectedChild in unexpectedChildrenWithNewNames {
+        precondition(unexpectedChild.isUnexpectedNodes)
+        let i = firstIndexOfChild(named: unexpectedChild.name)
+
+        let earlier = children[checked: i - 1]
+        let later = children[checked: i + 1]
+        precondition(!(earlier?.isUnexpectedNodes ?? false) && !(later?.isUnexpectedNodes ?? false))
+
+        let newChild = Child(forUnexpectedBetween: earlier, and: later, newerChildPath: [unexpectedChild])
+        precondition(newChild.name != unexpectedChild.name)
+        assert(!children.contains { $0.name == newChild.name })
```

Comment:
Any reason why this is `assert` and the other is `precondition`. I would prefer `precondition` so that we also hit it when building CodeGeneration in release mode.

Reply:
It just looked relatively expensive, but sure, I can change it.

Reply:
Since CodeGeneration is only run occasionally during swift-syntax development, I don’t really care about performance all that much.